cli_tools/gce_windows_upgrade/upgrader/upgrader.go

```diff
@@ -0,0 +1,458 @@
+//  Copyright 2020 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package upgrader
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+	"strings"
+
+	daisyutils "github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/daisy"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/param"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/common/utils/path"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/daisycommon"
+	"github.com/GoogleCloudPlatform/compute-image-tools/daisy"
+	daisyCompute "github.com/GoogleCloudPlatform/compute-image-tools/daisy/compute"
+	"google.golang.org/api/compute/v1"
+	"google.golang.org/api/option"
+)
+
+const (
+	logPrefix                      = "[windows-upgrade]"
+	rebootWorkflowPath             = "daisy_workflows/windows_upgrade/reboot.wf.json"
+	upgradePreparationWorkflowPath = "daisy_workflows/windows_upgrade/windows_upgrade_preparation.wf.json"
+	cleanupWorkflowPath            = "daisy_workflows/windows_upgrade/cleanup.wf.json"
+	rollbackWorkflowPath           = "daisy_workflows/windows_upgrade/rollback.wf.json"
+
+	rfc1035       = "[a-z]([-a-z0-9]*[a-z0-9])?"
+	projectRgxStr = "[a-z]([-.:a-z0-9]*[a-z0-9])?"
+
+	metadataKeyWindowsStartupScriptURL       = "windows-startup-script-url"
+	metadataKeyWindowsStartupScriptURLBackup = "windows-startup-script-url-backup"
+
+	versionWindows2008r2 = "windows-2008r2"
+	versionWindows2012r2 = "windows-2012r2"
+
+	upgradeIntroductionTemplate = "The following resources will be created/touched during the upgrade. " +
+		"Please record their name in order for cleanup or manual rollback.\n" +
+		"All resources are in project '%v', zone '%v'.\n" +
```

Comment:
This looks like a *fantastic* opportunity for using templates: https://golang.org/pkg/text/template/

Reply:
Looks not like I need such a rich-logic generation. The text here is pretty static.

Reply:
You can use ` instead of " to potentially make the text more readable.

Reply:
How are you confident that your current "%v" parameters receive the correct values (now, and in future iterations?)

Reply:
Those params are just for showing which resources are created/accessed. To make it clear, the text here contains 2 parts:
1. A resources list for manual cleanup/rollback.
2. guide of manual cleanup/rollback.

The #1 is as simple as a list.

The #2 actually will be a public doc URL. Right now, I put it in code as a placeholder. The reason is that before the solution is locked down, the guide can be modified. Will work with tech writer to provide the final guide later.